shared/utils/domains.ts

```diff
@@ -1,85 +1,64 @@
 // a large list of possible TLD's which increase the size of the bundle
 // unnecessarily for our usecase of trusted input.
-export function parseDomain(url?: string): Domain | null | undefined {
-  if (typeof url !== "string") {
-    return null;
-  }
-  if (url === "") {
-    return null;
+export function parseDomain(url: string): Domain {
+  if (!url) {
+    throw new TypeError("a non-empty url is required");
   }
 
-  // strip extermeties and whitespace from input
-  const normalizedDomain = trim(url.replace(/(https?:)?\/\//, ""));
-  const parts = normalizedDomain.split(".");
-
-  // ensure the last part only includes something that looks like a TLD
-  function cleanTLD(tld = "") {
-    return tld.split(/[/:?]/)[0];
-  }
+  const normalBaseUrl = normalizeUrl(env.URL);
+  const host = normalizeUrl(url);
 
-  // simplistic subdomain parse, we don't need to take into account subdomains
-  // with "." characters as these are not valid in Outline
-  if (parts.length >= 3) {
-    return {
-      subdomain: parts[0],
-      domain: parts[1],
-      tld: cleanTLD(parts.slice(2).join(".")),
-    };
-  }
+  // if the url doesn't include the base url, then it must be a custom domain
+  const baseUrlStart =
+    host === normalBaseUrl ? 0 : host.indexOf(`.${normalBaseUrl}`);
 
-  if (parts.length === 2) {
-    return {
-      subdomain: "",
-      domain: parts[0],
-      tld: cleanTLD(parts.slice(1).join(".")),
-    };
+  if (baseUrlStart === -1) {
+    return { teamSubdomain: "", host, custom: true };
   }
 
-  // one-part domain handler for things like localhost
-  if (parts.length === 1) {
-    return {
-      subdomain: "",
-      domain: cleanTLD(parts.slice(0).join()),
-      tld: "",
-    };
-  }
+  // we consider anything in front of the baseUrl to be the subdomain
+  const subdomain = host.substring(0, baseUrlStart);
+  const isReservedSubdomain = RESERVED_SUBDOMAINS.includes(subdomain);
 
-  return null;
+  return {
+    teamSubdomain: isReservedSubdomain ? "" : subdomain,
+    host,
+    custom: false,
+  };
 }
 
-export function stripSubdomain(hostname: string) {
-  const parsed = parseDomain(hostname);
-  if (!parsed) {
-    return hostname;
-  }
-  if (parsed.tld) {
-    return `${parsed.domain}.${parsed.tld}`;
-  }
-  return parsed.domain;
-}
+// export function isCustomDomain(hostname: string) {
+//   return parseDomain(hostname)?.custom ?? false;
+// }
 
-export function isCustomSubdomain(hostname: string) {
-  const parsed = parseDomain(hostname);
+export function getCookieDomain(domain: string) {
+  // always use the base URL for cookies when in hosted mode
+  // and the domain is not custom
+  if (env.SUBDOMAINS_ENABLED) {
+    const parsed = parseDomain(domain);
 
-  if (
-    !parsed ||
-    !parsed.subdomain ||
-    parsed.subdomain === "app" ||
-    parsed.subdomain === "www"
-  ) {
-    return false;
+    if (!parsed.custom) {
+      return normalizeUrl(env.URL);
```

Comment:
now that I look at it, I think this should also be `new URL(env.URL).host`

Reply:
Agreed, it should be `getoutline.com` in production

Reply:
What's the env URL -- is it http://getoutline.com ?

Reply:
It's `https://app.getoutline.com`

Reply:
Ah ok - need to change some assumptions then